django/core/handlers/asgi.py

```diff
@@ -189,6 +192,36 @@ async def handle(self, scope, receive, send):
         # Send the response.
         await self.send_response(response, send)
 
+    async def listen_for_disconnect(self, receive):
+        """Listens for http.disconnect from the client."""
+        while True:
+            message = await receive()
+            if message["type"] == "http.disconnect":
+                raise RequestAborted()
+
+    async def disconnect_or_start_request(self, scope, receive):
+        """Checks for client disconnect as the signal to start request is being sent."""
+        tasks = [
+            asyncio.ensure_future(coro)
+            for coro in (
+                self.signal_to_start_request(scope),
+                self.listen_for_disconnect(receive),
+            )
+        ]
+        done, pending = await asyncio.wait(tasks, return_when=asyncio.FIRST_COMPLETED)
+        try:
+            done.pop().result()
+        except RequestAborted:
+            raise RequestAborted()
+
+    async def signal_to_start_request(self, scope):
+        """Signals handler to start request."""
+        # Force self.listen_for_disconnect to run first
+        await asyncio.sleep(0)
```

Comment:
asyncio.wait is inconsistent with the task it runs first. forcing signal_to_start_request to await sleep allows listen_for_disconnect to run first (at least once).